server/lib/src/idm/server.rs

```diff
@@ -1275,127 +1276,133 @@ impl<'a> IdmServerAuthTransaction<'a> {
-                        softlock_write.commit();
-                        slock
-                    }
-                };
-                Some(slock_ref)
-            }
-            None => None,
+        let cred = if acp.allow_primary_cred_fallback() == Some(true) {
+            account
+                .unix_extn()
+                .and_then(|extn| extn.ucred())
+                .or_else(|| account.primary())
+        } else {
+            account.unix_extn().and_then(|extn| extn.ucred())
         };
 
-        let maybe_slock = if let Some(s) = maybe_slock_ref.as_ref() {
-            Some(s.lock().await)
-        } else {
-            None
+        let (cred, cred_id, cred_slock_policy) = match cred {
+            None => {
+                if acp.allow_primary_cred_fallback() == Some(true) {
+                    security_info!("Account does not have a posix or primary password configured.");
+                } else {
+                    security_info!("Account does not have a posix password configured.");
+                }
+                //TODO: Did we want to return an error condition?
+                return Ok(None);
+            }
+            Some(cred) => (cred, cred.uuid, cred.softlock_policy()),
         };
 
-        let maybe_valid = if let Some(mut slock) = maybe_slock {
-            // Apply the current time.
-            slock.apply_time_step(ct);
-            // Now check the results
-            if slock.is_valid() {
-                Some(slock)
+        let slock_ref = {
+            let softlock_read = self.softlocks.read();
+            if let Some(slock_ref) = softlock_read.get(&cred_id) {
+                slock_ref.clone()
             } else {
-                None
+                let _session_ticket = self.session_ticket.acquire().await;
+                let mut softlock_write = self.softlocks.write();
+                let slock = Arc::new(Mutex::new(CredSoftLock::new(cred_slock_policy)));
+                softlock_write.insert(cred_id, slock.clone());
+                softlock_write.commit();
+                slock
             }
-        } else {
-            None
         };
 
-        // Validate the unix_pw - this checks the account/cred lock states.
-        let res = if let Some(mut slock) = maybe_valid {
-            // Account is unlocked, can proceed.
-            account
-                .verify_unix_credential(uae.cleartext.as_str(), &self.async_tx, ct)
-                .inspect(|res| {
-                    if res.is_none() {
-                        // Update it.
-                        slock.record_failure(ct);
-                    };
-                })
-        } else {
-            // Account is slocked!
+        let mut slock = slock_ref.lock().await;
+
+        slock.apply_time_step(ct);
+
+        if !slock.is_valid() {
             security_info!("Account is softlocked.");
-            Ok(None)
+            //TODO: Did we want to return an error condition?
+            return Ok(None);
```

Comment:
ERr means "backend went bang" but Ok(None) indicates the auth failed and callers expect that.